rep-2008.rst

```diff
@@ -0,0 +1,597 @@
+                         |      |  |            +-------+------+
+                         |      |  |                    |
+                         +------+  |                    |
+                           LTTng   +--------------------+
+                                       re-instrument
+
+
+
+
+
+
+
+The following proposes a methodology to analyze a ROS 2 application and design appropriate acceleration:
+
+1. instrument both the core components of ROS 2 and the target kernels using `LTTng <https://lttng.org/>`_. Refer to `ros2_tracing <https://gitlab.com/ros-tracing/ros2_tracing>`_ for tools, documentation and ROS 2 core layers tracepoints;
+2. trace and benchmark the kernels on the CPU to establish a compute baseline;
+3. develop a hardware accelerated implementation on alternate hardware (e.g., GPU, FPGA, etc):
+
+   - **3.1** accelerate computations at the Node or Component level for each one of those identified in **2.** as good candidates.
+   - **3.2** accelerate inter-Node exchanges and reduce the overhead of the ROS 2 message-passing system across all its abstraction layers.
+
+4. trace, benchmark against the CPU baseline, and improve the accelerated implementation.
+
+The proposed ROS 2 methodology for hardware acceleration is demonstrated in [7]_ and [8]_.
+
+
+.. _acceleration examples:
+
+Acceleration examples
+^^^^^^^^^^^^^^^^^^^^^
+
+For the sake of illustrating maintainers and consumers how to build their own acceleration kernels and guarantee interoperability across technologies, a ROS 2 meta-package named `acceleration_examples <https://github.com/ros-acceleration/acceleration_examples>`_ will be maintained and made available. This meta-package will contain various packages with simple common acceleration examples. Each one of these examples should support all hardware acceleration solutions complying with this REP.
+
+In turn, a CI system will be set to build periodically and for every commit the meta-package.
+
+.. _capabilities:
+
+Capabilities
+^^^^^^^^^^^^
+
+The following list describes the hardware acceleration capabilities that hardware acceleration vendors must consider when connecting their firmware and technology solutions to the ROS 2 ecosystem.
+
+.. _Kernel Levels Capabilities:
+
+1. **Kernel Levels:**
+
+   .. _1.i:
+
+   i. Must provide at least one package producing `Level I kernels`_
+
+   .. _1.ii:
+
+   ii. Must provide at least one package producing `Level II kernels`_
+
+   .. _1.iii:
+
+   iii. Must provide at least one package producing `Level III kernels`_
+
+
+.. _Build System Enhancements:
+
+2. **Build System Enhancements:**
+
+   .. _2.i:
+
+   i. Must have an ``ament_<technology>`` (e.g. ``ament_vitis``) package that integrates technology-specific frameworks and software platforms into the ROS 2 build system through a series of CMake macros that maintainers can use to generate acceleration kernels from their `CMakeLists.txt` files.
```

Comment:
I think the requirement these two  requirements (i  and ii) should be optional.
I am not familiar with cuda, but I think it is now supported by new versions of cmake, so I think one can do without ament_cuda for example.

For GPUs, with opencl, you can runtime compile/link programs. In this case as well there is no need for ament_<technology>


This comment is to be taken with a grain of salt

Reply:
What we're trying to achieve in here is that the same package meta-data (i.e. the `CMakeLists.txt`) contains logic that allows producing acceleration kernels for various accelerators without having to modify it. This is achieved by integrating acceleration frameworks into `ament` build system via CMake macros that abstract away the specifics of each technology. E.g.:

- `ament_acceleration`
  - `ament_vitis`
  - `ament_cuda`
  - `ament_<whatever>`

If newer versions of CUDA can be easily integrated with CMake directly that's fantastic, but we still need the "*ROS build system alignment*". That's what this requirement in here captures. I think this is important.